### PR TITLE
IOPath interface and IOPathSpec implementation

### PR DESCRIPTION
**Pull request by cmnbroad**
An interface and extensible concrete implementation for paths specifiers backed by URIs. The hard part is the names.

### CODE REVIEW

**Comment by yfarjoun** [comment]
@cmnbroad could you elaborate a little on what problem this solves? what will be easier to do using this interface?

**Comment by cmnbroad** [comment]
Yeah, of course. The idea is to have a single interface and concrete implementation for use in existing and new htsjdk interfaces, for specifying input or output resources that are on a local file system; a remote file system such as GCS, or some other resource; such as an htsget or refget uri/specifier.

The constructor accepts a string that can be either an absolute or relative local file name, or a URI, including scheme and query parameters. The (IOPaths) are always backed by a URI with a protocol scheme (`file://` if nothing else) that can be queried to determine how to handle the resource, even if one wasn't provided to the constructor. The idea is to provide a single currency (for new code at least), and help eliminate the proliferation of `File`/`Path`/`String` method overloads and interconversions, and eliminate ambiguities around handling path names that contain embedded characters such as space or '#' (like [this one](https://github.com/samtools/htsjdk/pull/1487)), since these are always backed by a normalized, encoded URI.

Callers can provide subclassed implementations for resolving paths that require special handling (GATK for example uses this to provide to a `Path` for both gs:// and hdfs:// access).

This will be the first step toward support for versioned, side-by-side file formats and https://github.com/samtools/htsjdk/issues/1482, though I think it will be useful even with the existing classes.

File: src/main/java/htsjdk/io/IOPath.java [hunk 1]
```diff
@@ -0,0 +1,192 @@
+package htsjdk.io;
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply src/main/java/htsjdk/io/IOPath.java:1]: Yeah, the first commit has the GATK versions. The only changes were the package and class names, exceptions, and a handful of test result changes to reflect the different file systems available in htjsjdk.

File: src/main/java/htsjdk/io/IOPathSpec.java [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+ * More specifically:
+ *
+ *     absoluteURI   = scheme ":" ( hier_part | opaque_part )
+ *         hier_part     = ( net_path | abs_path ) [ "?" query ]
+ *         net_path      = "//" authority [ abs_path ]
+ *         abs_path      = "/"  path_segments
+ *         opaque_part   = uric_no_slash *uric
+ *         uric_no_slash = unreserved | escaped | ";" | "?" | ":" | "@" | "&" | "=" | "+" | "$" | ","
+ */
+public class IOPathSpec implements IOPath, Serializable {
```
[the remark this answers is not in the document]
Comment (cmnbroad) [reply src/main/java/htsjdk/io/IOPathSpec.java:70]: HtsjdkPath is fine with me.

**Comment by cmnbroad** [comment]
I updated the class name to `HtsjdkPath`, leaving the interface name as IOPath, and changed some test variable/method/provider names to reflect that change.

**Comment by cmnbroad** [comment]
Any other community comments/input before we move forward with this ?

**Comment by lbergelson** [comment]
No one is chiming in so lets merge this.

**Pull request merged**